flow/layers/backdrop_filter_layer.cc

```diff
@@ -11,13 +11,57 @@ BackdropFilterLayer::BackdropFilterLayer(sk_sp<SkImageFilter> filter)
 
 BackdropFilterLayer::~BackdropFilterLayer() = default;
 
+void BackdropFilterLayer::initRetained(std::shared_ptr<Layer> retainedLayer) {
+  BackdropFilterLayer* bdfLayer = (BackdropFilterLayer*)retainedLayer.get();
+  if (bdfLayer->retainedBackdrop_) {
+    retainedBackdrop_ = bdfLayer->retainedBackdrop_;
+  } else {
+    retainedBackdrop_ = std::move(retainedLayer);
+  }
+}
+
+void BackdropFilterLayer::Preroll(PrerollContext* context,
+                                  const SkMatrix& matrix) {
+  ContainerLayer::Preroll(context, matrix);
+
+  if (retainedBackdrop_ && context->raster_cache &&
+      context->view_embedder == nullptr) {
+    context->raster_cache->PrepareForSnapshot(context, retainedBackdrop_.get());
```

Comment:
Why not put this inside `BackdropFilterLayer::Paint` like
```
  if (retainedBackdrop_ && context.surface &&
      context.view_embedder == nullptr) {
    RasterCacheResult my_cache = context->raster_cache->PrepareOrGetSnapshot(...)
```

Reply:
There is a distinct way that the cache is accessed in the preroll-compatible methods and the paint-compatible methods. The paint-compatible methods are geared towards a static cache that has all of its keys already created.

So, I went with this design to better match the usage patterns in the other caching layers.

Reply:
I see. Nevertheless, we're already changing the usage patterns (previously Paint only has const access to the raster cache), so please feel free to break the old pattern to make the code a little more shorter and cleaner. (Another way to interpret our pattern is to only create the key when it's about to snapshot a raster cache; previously, they happened in Preroll; now, they can also happen in Paint.)

Reply:
In the version of this code in my latest commit I got rid of the Preroll code and now this all happens in the Paint method instead.